File: udemy11imp/non_RepeatingCharacter.java
import java.util.HashMap;

public class nr{
	public static void main(String[] args){
		System.out.println(nonRepeating("abcab")); // should return 'c'
        System.out.println(nonRepeating("abab")); // should return null
        System.out.println(nonRepeating("aabbbc")); // should return 'c'
        System.out.println(nonRepeating("aabbdbc")); // should return 'd'
	}

	public static Character nonRepeating(String s){
		HashMap<Character, Integer> Counter = new HashMap<Character, Integer>();
		for (char c: s.toCharArray()){
			if (Counter.containsKey(c)){
				Integer newVal = Counter.get(c)+1;
				Counter.put(c, newVal);
			} else {
				Counter.put(c, 1);
			}
		}
		for (char c: s.toCharArray()){
			if (Counter.get(c) == 1) return c;
		}
		return null;
	}
}

File: udemy11imp/non_repeatingCharacter.py
def non_repeating(string):
  counter = {}
  for c in string:
    if c not in counter:
      counter[c] = 1
    else:
      counter[c]+=1
  for c in string:
    if counter[c]==1:
      return c
  return None
  
  
print(non_repeating("abcab")) # should return 'c'
print(non_repeating("abab")) # should return None
print(non_repeating("aabbbc")) # should return 'c'
print(non_repeating("aabbdbc")) # should return 'd'
